jose/validate.go

```diff
@@ -49,8 +50,33 @@ func validateX5(certs []*x509.Certificate, key interface{}) error {
 		return errors.New("certs cannot be empty")
 	}
 
-	if err := keyutil.VerifyPair(certs[0].PublicKey, key); err != nil {
-		return errors.Wrap(err, "error verifying certificate and key")
+	// Compare public keys if we have an opaque signer, otherwise check that private/public match
+	if opaqueSigner, isOpaqueSigner := key.(OpaqueSigner); isOpaqueSigner {
+		signerPub, ok := opaqueSigner.Public().Key.(crypto.PublicKey)
+		var publicKeysMatch = true
+		if !ok {
+			return errors.Errorf("opaqueSigner public key type %T is not supported", signerPub)
+		}
+
+		//crypto.PublicKey is actually an empty interface?! So we have to explicitly check each type?! Annoying!
+		switch pub := certs[0].PublicKey.(type) {
+		case *rsa.PublicKey:
+			publicKeysMatch = pub.Equal(signerPub)
+		case *ecdsa.PublicKey:
+			publicKeysMatch = pub.Equal(signerPub)
+		case ed25519.PublicKey:
+			publicKeysMatch = pub.Equal(signerPub)
+		default:
+			return errors.Errorf("unsupported public key type %T", pub)
+		}
```

Comment:
PublicKey is an empty interface, but you should be able to create a private interface and test against it. 

You can even an anonymous one:

```go
pub, ok := certs[0].PublicKey.((interface{ Equal(crypto.PublicKey) bool })
if ok {
	publicKeysMatch = pub.Equal(signerPub)
} else {
	return errors.Errorf("unsupported public key type %T", certs[0].PublicKey)
}
```

Reply:
@unreality take a look at the linter errors too.

Reply:
> PublicKey is an empty interface, but you should be able to create a private interface and test against it.

Neat, thanks for the tip :)

Reply:
@unreality do you want to do this change?

Reply:
@maraino i thought it was done in commit [6afc19d](https://github.com/smallstep/crypto/pull/105/commits/6afc19dfc9c4b38250c1572d00beafef94550270) ?

Have i messed up? :)

Reply:
I messed up, I clicked on the [jose/validate.go](https://github.com/smallstep/crypto/pull/105/files/ceedd033352f40cfff1455eb1f28d5fd92610990#diff-019ea4b9c6a1e4a6a5d2054d144cc4c9db037800ce6f3aae0efe90480596d862) of this comment and showed me the old version.